### DISCUSSION TITLE
`ADR Suggestion` EasyScience Organization Repositories

### DISCUSSION THREAD

**Discussion by AndrewSazonov** [easyscience/.github#20] [Ideas]
# Proposal to Rename EasyScience Organization Repositories

The EasyScience organization currently includes 50 repositories. It has become challenging to navigate and track them efficiently.

To address this, I propose a systematic renaming repositories to improve clarity and usability. The key changes include:

1. **Removing the "Easy" Prefix**:
   - Repository names like "EasyDiffractionLib" will be shortened to "diffraction-lib" to reduce redundancy and improve readability.
   - The "Easy" prefix is unnecessary as all repositories are under the EasyScience organization, which already establishes their association.
   - Note: This suggestion applies **only to repository names**, not project names. For instance, the project will still be called "EasyDiffraction," and Python packages will retain their names, such as `easydiffraction`, `easyreflectometry`, `easyscience`, etc.

2. **Grouping Repositories by Categories**:
   - Repositories are organized into logical groups (e.g., specific projects, generic projects, common resources, third-party dependencies) to make navigation easier.

Below are the proposed changes for repository names, categorized for clarity. The first column lists the new repository name, and the second column provides a description to be aligned with the "About" section of the repository.

Note: QENS is not mentioned yet because the final decision on the name of the project has not yet been made.

---

### EasyDiffraction

| Repository             | Description                                                    |   
|------------------------|----------------------------------------------------------------|
| [diffraction]          | Launching point for the EasyDiffraction family of repositories |
| [diffraction-lib]      | Python library for diffraction data analysis                   |
| [diffraction-app]      | Desktop application for diffraction data analysis              |
| [diffraction-page]     | Main web page for EasyDiffraction                              |
| [diffraction-lib-docs] | User documentation for the diffraction library                 |
| [diffraction-app-docs] | User documentation for the diffraction application             |

### EasyReflectometry

| Repository           | Description                                                      |
|----------------------|------------------------------------------------------------------|
| [reflectometry]      | Launching point for the EasyReflectometry family of repositories |
| [reflectometry-lib]  | Python library for reflectometry data analysis                   |
| [reflectometry-app]  | Desktop application for reflectometry data analysis              |
| [reflectometry-page] | Main web page for EasyReflectometry                              |

### EasyImaging

| Repository    | Description                                                |
|---------------|------------------------------------------------------------|
| [imaging]     | Launching point for the EasyImaging family of repositories |
| [imaging-lib] | Python library for Bragg-edge imaging data analysis        |
| [imaging-app] | Desktop application for Bragg-edge imaging data analysis   |

### EasyTexture

| Repository    | Description                                    |
|---------------|------------------------------------------------|
| [texture-app] | Desktop application for texture data reduction |

---

## Generic Projects

| Repository            | Description                                                                |
|-----------------------|----------------------------------------------------------------------------|
| [.github]             | EasyScience organization profile                                           |
| [easyscience-core]    | Core building blocks for Python libraries in the EasyScience framework     |
| [easyscience-app]     | Graphical components for desktop applications in the EasyScience framework |
| [easyscience-crystal] | Backend crystallographic tools in the EasyScience framework                |
| [easyscience-page]    | Main web page for the EasyScience organization                             |

---

## Common Resources

| Repository          | Description                                                   |
|---------------------|---------------------------------------------------------------|
| [assets-branding]   | Logos, icons, and branding assets for the entire organization |
| [assets-page]       | Shared assets for project web pages                           |
| [assets-docs]       | Shared assets for project documentation                       |
| [templates-project] | Copier template for technique-specific EasyScience projects   |

---

## Other EasyScience Repositories

| Repository         | Description                                                 |
|--------------------|-------------------------------------------------------------|
| [pypi]             | Custom Python package index for EasyScience-based libraries |

---

## Third-Party Projects

| Repository               | Description                                  |
|--------------------------|----------------------------------------------|
| [dependencies-cryspy]    | Custom version of the CrysPy library         |
| [dependencies-pdffit2]   | Custom version of the diffpy.pdffit2 library |
| [dependencies-pycrysfml] | Python bindings for the CrysFML2008 library  |

[diffraction]: https://github.com/EasyScience/EasyDiffraction
[diffraction-lib]: https://github.com/EasyScience/EasyDiffractionLib
[diffraction-app]: https://github.com/EasyScience/EasyDiffractionApp
[diffraction-page]: https://easyscience.github.io/EasyDiffractionWww
[diffraction-lib-docs]: https://easyscience.github.io/EasyDiffractionLibDocs
[diffraction-app-docs]: https://easyscience.github.io/EasyDiffractionAppDocs

[reflectometry]: https://github.com/EasyScience/EasyReflectometry
[reflectometry-lib]: https://github.com/EasyScience/EasyReflectometryLib
[reflectometry-app]: https://github.com/EasyScience/EasyReflectometryApp
[reflectometry-page]: https://easyscience.github.io/EasyReflectometryWww

[texture-app]: https://github.com/EasyScience/EasyTextureApp

[imaging]: https://github.com/EasyScience/EasyReflectometry
[imaging-lib]: https://github.com/EasyScience/EasyImagingLib
[imaging-app]: https://github.com/EasyScience/EasyImagingApp

[.github]: https://github.com/EasyScience/.github
[easyscience-core]: https://github.com/EasyScience/EasyScience
[easyscience-app]: https://github.com/EasyScience/EasyApp
[easyscience-crystal]: https://github.com/EasyScience/EasyCrystallography
[easyscience-page]: https://github.com/EasyScience/easyScienceWww

[assets-branding]: https://github.com/EasyScience/BrandingResources
[assets-page]: https://github.com/EasyScience/EasySite
[assets-docs]: https://github.com/EasyScience/CommonDocsAssets
[templates-project]: https://github.com/EasyScience/EasyProjectTemplate

[pypi]: https://github.com/EasyScience/pypi

[dependencies-cryspy]: https://github.com/EasyScience/cryspy
[dependencies-pycrysfml]: https://github.com/EasyScience/PyCrysFML
[dependencies-pdffit2]: https://github.com/EasyScience/pdffit2

**Comment by rozyczko** [comment]
This looks very good. I wonder if it weren't even more structured if we had a directory structure in our repo, like

```
 ├───EasyScience
 |   └───Diffraction
 |          └───Lib
 |          └───App
 |          └───Docs
 |          └───Www
 |   └───Reflectometry
 |          └───Lib
 |          └───App
 |          └───Docs
 |          └───Www
 |   └───Common
 ...
```

This can be achieved with either submodules or a monorepo.

**Reply by AndrewSazonov** [reply]
**Monorepo** and **submodules** seem like interesting alternatives to the current approach. However, while both approaches offer advantages, they also introduce complexities that may outweigh their benefits in our specific case.

**Monorepo** challenges:
- Increased Build Complexity: 
  - Our existing build scripts, which already include conditional logic, would require even more granular checks. For instance:
    - Modifications to the main web page should not trigger builds for the library, app, or their respective documentation.
    - Changes to the app would often not require building the library or its documentation. 
  - Managing these distinctions in a monorepo would demand more complex CI/CD workflows.
- Workflow Management:
  - Currently, each repository has its own workflows visible on the GitHub Actions page. Combining multiple repositories into one would significantly increase the number of workflows in a single list.
  - Naming workflows to indicate their corresponding sub-project (e.g., “diffraction-app-build”) would add complexity and reduce readability.
  - Filtering workflows by branch or sub-project would also become more challenging.
- Release Management:
  - The library and application have separate release cycles. A monorepo would make tagging releases problematic, as it would be unclear whether a tag corresponds to a library release, an app release, or both.
  - Maintaining independent versioning for different components in a monorepo can be cumbersome.
- Merge Conflicts: 
  - The likelihood of conflicts would increase, especially when multiple contributors work on different parts of the repository simultaneously. This could slow down development.
- Repository Size: 
  - Over time, a monorepo could become very large, making operations like cloning or searching slower.

**Submodules** challenges:
- Complex Development Workflow:  
   - Developers must initialize and update submodules explicitly when cloning or pulling changes. This adds friction, especially for new contributors.
   - Submodules track specific commits, not branches, requiring extra effort to keep them in sync with the latest changes.
- CI/CD Pipeline Complexity: 
  - Integrating submodules into CI/CD pipelines requires additional configuration to ensure the correct versions of submodules are checked out and built together.
- No Reduction in Repositories: 
  - Unlike a monorepo, submodules do not consolidate repositories. While they improve organization, they do not simplify repository management.

While both the **monorepo** and **submodules** approaches have potential advantages, they introduce significant complexities in terms of workflow, build pipelines, and release management. Given these considerations, I suggest adopting the simpler solution of renaming and reorganizing the existing repositories as proposed in the original comment.

By starting with a simpler renaming strategy, we can achieve immediate improvements in organization and usability while avoiding the challenges and overhead of more complex solutions. Over time, if we find additional structure beneficial, we can reconsider options like a monorepo or submodules.

**Comment by rozyczko** [comment]
Some comments on the common modules names

`EasyCrystallography` -> `easyscience-crystal`: `crystal` sounds more like a library for single crystal to me, maybe the full name `easyscience-crystallography` or `easyscience-symmetry`? 

`EasyApp` -> `easyscience-app`: `-app` does not immediately convey that this is a GUI-related library. Maybe `easyscience-gui` or `easyscience-guilib` (this is a library, not a standalone app).

`EasyScience` -> `easyscience-core`: `-core` - maybe we should, again, add `-lib` to signify this is a component `easyscience-corelib` or `easyscience-core-lib`

**Reply by AndrewSazonov** [reply]
Okay, let’s apply these changes for better clarity and consistency:

- EasyCrystallography → easyscience-crystallography
- EasyApp → easyscience-gui-lib
- EasyScience → easyscience-core-lib

**Comment by damskii9992** [comment]
Writing the comment from our daily discussion here:
Renaming the repositories might interfere with users finding our github repositories when googling for them as they won't be listed with their project names.

**Reply by AndrewSazonov** [reply]
While renaming repositories might initially affect search visibility, the impact can be minimized by optimizing the README files, as their content is indexed by Google and contributes significantly to search rankings. Therefore, including the project’s full name (e.g., “EasyDiffraction”) in the README file will help maintain visibility even after a repository is renamed. Additionally, descriptions, tags, topics, and other metadata associated with the repository are indexed and also contribute to search rankings.

**Comment by AndrewSazonov** [comment]
Here is the finalized version of the EasyScience organization README, updated to reflect the changes discussed during the review process:

<p align='left'>
  <picture>
    
    <source media='(prefers-color-scheme: light)' srcset='https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyScience/logos/es-logo_light.svg'>
    
    <source media='(prefers-color-scheme: dark)' srcset='https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyScience/logos/es-logo_dark.svg'>
    
    <img src='https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyScience/logos/es-logo_light.svg' alt='EasyCrystallography'>
  </picture>
</p>

**EasyScience** is a versatile data analysis framework designed to streamline the journey from data collection to publication. It provides:
- **Powerful back-end tools**: Core functionality through Python libraries.
- **Efficient front-end modules**: QML components for developing desktop applications with ease.

Explore our dedicated repositories for specific techniques:

|                             | 🏠 Project Hub      | 📦 Python Library      | 💻 Desktop Application | 🌐 Main Web Page       |
|-----------------------------|---------------------|------------------------|------------------------|------------------------|
| ![ed-dark]![ed-light]       | [diffraction]       | [diffraction-lib]      | [diffraction-app]      | [diffraction-page]     |
| ![er-dark]![er-light]       | [reflectometry]     | [reflectometry-lib]    | [reflectometry-app]    | [reflectometry-page]   |
| ![ei-dark]![ei-light]       | [imaging]           | [imaging-lib]          | [imaging-app]          |                        |
| ![eq-dark]![eq-light]       | [dynamics]          | [dynamics-lib]         | [dynamics-app]         |                        |
| ![et-dark]![et-light]       |                     |                        | [texture-app]          |                        |
| ![esh-dark]![esh-light]     |                     |                        | [shapes-app]           |                        |

## Full List of Repositories

### 1. Technique-Specific Projects

|                             | Repository              | Description                                                                                         |
|-----------------------------|-------------------------|-----------------------------------------------------------------------------------------------------|
| ![ed-dark]![ed-light]       | 🏠 [diffraction]        | Launching point for the EasyDiffraction family of repositories                                      |
|                             | 📦 [diffraction-lib]    | Python library for diffraction data analysis                                                        |
|                             | 💻 [diffraction-app]    | Desktop application for diffraction data analysis                                                   |
|                             | 🌐 [diffraction-page]   | Main web page for EasyDiffraction                                                                   |
| ![er-dark]![er-light]       | 🏠 [reflectometry]      | Launching point for the EasyReflectometry family of repositories                                    |
|                             | 📦 [reflectometry-lib]  | Python library for reflectometry data analysis                                                      |
|                             | 💻 [reflectometry-app]  | Desktop application for reflectometry data analysis                                                 |
|                             | 🌐 [reflectometry-page] | Main web page for EasyReflectometry                                                                 |
| ![ei-dark]![ei-light]       | 🏠 [imaging]            | Launching point for the EasyImaging (Bragg-edge imaging) family of repositories                     |
|                             | 📦 [imaging-lib]        | Python library for Bragg-edge imaging data analysis                                                 |
|                             | 💻 [imaging-app]        | Desktop application for Bragg-edge imaging data analysis                                            |
| ![eq-dark]![eq-light]       | 🏠 [dynamics]           | Launching point for the EasyDynamics (Quasielastic Neutron Scattering, QENS) family of repositories |
|                             | 📦 [dynamics-lib]       | Python library for quasielastic neutron scattering data analysis                                    |
|                             | 💻 [dynamics-app]       | Desktop application for quasielastic neutron scattering imaging data analysis                       |
| ![esh-dark]![esh-light] | 💻 [shapes-app]         | Desktop application for ???                                                                         |
| ![et-dark]![et-light]       | 💻 [texture-app]        | Desktop application for texture data reduction                                                      |

### 2. Generic Projects

|                           | Repository                    | Description                                                                |
|---------------------------|-------------------------------|----------------------------------------------------------------------------|
| ![es-dark]![es-light]     | [.github]                     | EasyScience organization profile                                           |
|                           | [easyscience-page]            | Main web page for the EasyScience organization                             |
|                           | [easyscience-corelib]         | Core building blocks for Python libraries in the EasyScience framework     |
| ![escr-dark]![escr-light] | [easyscience-crystallography] | Backend crystallographic tools in the EasyScience framework                |
|                           | [easyscience-guilib]          | Graphical components for desktop applications in the EasyScience framework |

### 3. Common Resources

| Repository          | Description                                                   |
|---------------------|---------------------------------------------------------------|
| [assets-branding]   | Logos, icons, and branding assets for the entire organization |
| [assets-page]       | Shared assets for project web pages                           |
| [assets-docs]       | Shared assets for project documentation                       |
| [templates-project] | Copier template for technique-specific EasyScience projects   |

### 4. Other EasyScience Repositories

| Repository | Description                                                 |
|------------|-------------------------------------------------------------|
| [pypi]     | Custom Python package index for EasyScience-based libraries |

### 5. Third-Party Projects

| Repository       | Description                                  |
|------------------|----------------------------------------------|
| [deps-cryspy]    | Custom version of the CrysPy library         |
| [deps-pycrysfml] | Python bindings for the CrysFML2008 library  |
| [deps-pdffit2]   | Custom version of the diffpy.pdffit2 library |


## Repositories to be Merged

Add the prefix `to_be_merged__` to repositories that will be integrated into others:

- [EasyDiffractionLibDocs](https://github.com/EasyScience/EasyDiffractionLibDocs): Will be merged into the `docs/` directory of [EasyDiffractionLib](https://github.com/EasyScience/EasyDiffractionLib).
- [EasyDiffractionAppDocs](https://github.com/EasyScience/EasyDiffractionAppDocs): Will be merged into the `docs/` directory of [EasyDiffractionApp](https://github.com/EasyScience/EasyDiffractionApp).

## Repositories to be Checked

Add the prefix `to_be_checked__` to repositories under review:

- [easyAppWww](https://github.com/EasyScience/easyAppWww)
- [easyCoreWww](https://github.com/EasyScience/easyCoreWww)
- [easyExampleLib](https://github.com/EasyScience/easyExampleLib)
- [easySkeletonApp](https://github.com/EasyScience/easySkeletonApp)
- [easyTemplateLib](https://github.com/EasyScience/easyTemplateLib)

## Repositories to be Removed

Add the prefix `to_be_removed__` to repositories slated for removal:

- [demo-repository](https://github.com/EasyScience/demo-repository)
- [diffpy.structure](https://github.com/EasyScience/diffpy.structure)
- [easyAppGui](https://github.com/EasyScience/easyAppGui)
- [easyAppLogic](https://github.com/EasyScience/easyAppLogic)
- [libsDarwin](https://github.com/EasyScience/libsDarwin)
- [libsLinux](https://github.com/EasyScience/libsLinux)
- [libsWin32](https://github.com/EasyScience/libsWin32)
- [matplotlib_backend_qtquick](https://github.com/EasyScience/matplotlib_backend_qtquick)
- [pycrysfml08](https://github.com/EasyScience/pycrysfml08)
- [EasyScience_old](https://github.com/EasyScience/EasyScience_old)
- [test_conditional](https://github.com/EasyScience/test_conditional)




[ed-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyDiffraction/logos/ed-logo_102x32_dark.svg#gh-dark-mode-only
[ed-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyDiffraction/logos/ed-logo_102x32_light.svg#gh-light-mode-only
[er-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyReflectometry/logos/er-logo_120x32_dark.svg#gh-dark-mode-only
[er-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyReflectometry/logos/er-logo_120x32_light.svg#gh-light-mode-only
[ei-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyImaging/logos/ei-logo_84x32_dark.svg#gh-dark-mode-only
[ei-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyImaging/logos/ei-logo_84x32_light.svg#gh-light-mode-only
[eq-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyDynamics/logos/eq-logo_93x32_dark.svg#gh-dark-mode-only
[eq-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyDynamics/logos/eq-logo_93x32_light.svg#gh-light-mode-only
[esh-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyShapes/logos/esh-logo_79x32_dark.svg#gh-dark-mode-only
[esh-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyShapes/logos/esh-logo_79x32_light.svg#gh-light-mode-only
[et-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyTexture/logos/et-logo_80x32_dark.svg#gh-dark-mode-only
[et-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyTexture/logos/et-logo_80x32_light.svg#gh-light-mode-only

[es-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyScience/logos/es-logo_81x32_dark.svg#gh-dark-mode-only
[es-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyScience/logos/es-logo_81x32_light.svg#gh-light-mode-only
[esco-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyDiffraction/logos/ed-logo_102x32_dark.svg#gh-dark-mode-only
[esco-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyDiffraction/logos/ed-logo_102x32_light.svg#gh-light-mode-only
[escr-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyCrystallography/logos/ecr-logo_132x32_dark.svg#gh-dark-mode-only
[escr-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyCrystallography/logos/ecr-logo_132x32_light.svg#gh-light-mode-only
[esap-dark]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyDiffraction/logos/ed-logo_102x32_dark.svg#gh-dark-mode-only
[esap-light]: https://raw.githubusercontent.com/EasyScience/BrandingResources/refs/heads/master/EasyDiffraction/logos/ed-logo_102x32_light.svg#gh-light-mode-only


[diffraction]: https://github.com/EasyScience/EasyDiffraction
[diffraction-lib]: https://github.com/EasyScience/EasyDiffractionLib
[diffraction-app]: https://github.com/EasyScience/EasyDiffractionApp
[diffraction-page]: https://easyscience.github.io/EasyDiffractionWww
[diffraction-lib-docs]: https://easyscience.github.io/EasyDiffractionLibDocs
[diffraction-app-docs]: https://easyscience.github.io/EasyDiffractionAppDocs


[reflectometry]: https://github.com/EasyScience/EasyReflectometry
[reflectometry-lib]: https://github.com/EasyScience/EasyReflectometryLib
[reflectometry-app]: https://github.com/EasyScience/EasyReflectometryApp
[reflectometry-page]: https://easyscience.github.io/EasyReflectometryWww


[imaging]: https://github.com/EasyScience/EasyReflectometry
[imaging-lib]: https://github.com/EasyScience/EasyImagingLib
[imaging-app]: https://github.com/EasyScience/EasyImagingApp


[dynamics]: https://github.com/EasyScience/EasyQens
[dynamics-lib]: https://github.com/EasyScience/EasyQensLib
[dynamics-app]: https://github.com/EasyScience/EasyQensApp


[shapes-app]: https://github.com/easyscience/shapes-app


[texture-app]: https://github.com/EasyScience/EasyTextureApp


[shape-app]: https://github.com/EasyScience/shape-app


[.github]: https://github.com/EasyScience/.github
[easyscience-corelib]: https://github.com/EasyScience/EasyScience
[easyscience-guilib]: https://github.com/EasyScience/EasyApp
[easyscience-crystallography]: https://github.com/EasyScience/EasyCrystallography
[easyscience-page]: https://github.com/EasyScience/easyScienceWww


[assets-branding]: https://github.com/EasyScience/BrandingResources
[assets-page]: https://github.com/EasyScience/EasySite
[assets-docs]: https://github.com/EasyScience/CommonDocsAssets


[templates-project]: https://github.com/EasyScience/EasyProjectTemplate
[pypi]: https://github.com/EasyScience/pypi


[deps-cryspy]: https://github.com/EasyScience/cryspy
[deps-pycrysfml]: https://github.com/EasyScience/PyCrysFML
[deps-pdffit2]: https://github.com/EasyScience/pdffit2

**Comment by rozyczko** [comment]
The last comment should be made a final ADR and the list of changes converted into issues in corresponding project repos.
I think we all agreed to the above names?
